hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java

```diff
@@ -371,27 +343,44 @@ public void deleteByControllerID(final String controllerID) {
     }
 
     @Override
-    public Page<Target> findByAssignedDistributionSet(final Pageable pageReq, final long distributionSetID) {
-        throwEntityNotFoundIfDsDoesNotExist(distributionSetID);
+    public Page<Target> findByAssignedDistributionSet(final Pageable pageReq, final long distributionSetId) {
+        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
+        return targetRepository.findByAssignedDistributionSetId(pageReq, distributionSetId);
+    }
 
-        return targetRepository.findByAssignedDistributionSetId(pageReq, distributionSetID);
+    @Override
+    public Page<Target> findByAssignedDistributionSetAndQuery(final Pageable pageable, final long distributionSetId,
+            final String query) {
+        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
+        final Set<String> controllerIds = targetRepository.findControllerIdsByAssignedDistributionSet(distributionSetId);
+        return targetQueryExecutionManagement.findByQuery(pageable, query, controllerIds);
     }
 
+    /*
+     * (non-Javadoc) Used by Auto-Assign checker: Identify new targets which have
+     * not an specific DS assigned.
+     */
     @Override
-    public Page<Target> findByAssignedDistributionSetAndRsql(final Pageable pageReq, final long distributionSetID,
-            final String rsqlParam) {
-        throwEntityNotFoundIfDsDoesNotExist(distributionSetID);
+    public Page<Target> findByNoActionWithDistributionSetExistsAndQuery(final Pageable pageable,
+            final long distributionSetId, final String query) {
+        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
+        final Set<String> controllerIds = targetRepository.findByNoActionWithDistributionSetExisits(distributionSetId);
```

Comment:
Yes that is one consequence of having the search external also might be a consequence in some micro-service architectures.
I think we can get rid by some of these "domain" dependend finder-methods when we just move the domain-join into the query. I.e. the syntax of the query must contain the "andByTag". Propably we need to modify or create query-strings - than it will be only one call as before.

Reply:
Looking forward to your proposal

Reply:
in this PR?

Reply:
Not necessarily. But I'd love to see this addressed before we merge to mainline